ImageBufferShareableBitmapBackend lacks ownership attribution

kkinnunen-apple opened the pull request:
#### cab1c34e3f5de8eb66ddab429127b3127cd33f65
<pre>
ImageBufferShareableBitmapBackend lacks ownership attribution
<a href="https://bugs.webkit.org/show_bug.cgi?id=267266">https://bugs.webkit.org/show_bug.cgi?id=267266</a>
<a href="https://rdar.apple.com/problem/120707603">rdar://problem/120707603</a>

Reviewed by Matt Woodrow.

Add ShareableBitmap::setOwnershipOfMemory() to attribute the
ShareableBitmap to the owner process, currently always WP.

Use this when creating RemoteImageBuffer instances.
Use ShareableBitmap::Handle::takeOwnershipOfMemory() when receiving
the handles in WP.

Make some call sites of ShareableBitmap::create() more consistent.
The general pattern for marking the owner has to be:
- Use ShareableBitmap::setOwnershipOfMemory() in GPUP side immediately
  after creation.
- Use ShareableBitmap::Handle::takeOwnershipOfMemory() in WP side
  when receiving the handle. This is needed for the older macOS systems.

Remove unused ImageBuffer::setOwnershipIdentity() and similar methods.
Currently the ImageBuffer API contract is that the creator owns the
memory, and the internal implementation, in this case calls to
ShareableBitmap ensure it.

* Source/WebCore/platform/graphics/ImageBufferBackend.h:
(WebCore::ImageBufferBackend::setOwnershipIdentity): Deleted.
* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.cpp:
(WebCore::ImageBufferIOSurfaceBackend::transferToNewContext):
* Source/WebKit/GPUProcess/graphics/RemoteImageBuffer.cpp:
(WebKit::RemoteImageBuffer::getShareableBitmap):
(WebKit::RemoteImageBuffer::filteredNativeImage):
* Source/WebKit/Platform/SharedMemory.h:
* Source/WebKit/Shared/ShareableBitmap.h:
* Source/WebKit/Shared/ShareableBitmapHandle.cpp:
(WebKit::ShareableBitmapHandle::setOwnershipOfMemory const):
* Source/WebKit/Shared/cairo/ShareableBitmapCairo.cpp:
(WebKit::ShareableBitmap::setOwnershipIdentity):
* Source/WebKit/Shared/cg/ShareableBitmapCG.mm:
(WebKit::ShareableBitmap::setOwnershipIdentity):
* Source/WebKit/WebProcess/GPU/graphics/ImageBufferShareableBitmapBackend.cpp:
(WebKit::ImageBufferShareableBitmapBackend::create):
(WebKit::ImageBufferShareableBitmapBackend::transferToNewContext):
(WebKit::BufferShareableBitmapBackend::bytesPerRow const):
(WebKit::ImageBufferShareableBitmapBackend::bytesPerRow const): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/ImageBufferShareableBitmapBackend.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.cpp:
(WebKit::RemoteImageBufferProxy::didCreateBackend):
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferShareableMappedIOSurfaceBitmapBackend.cpp:
(WebKit::ImageBufferShareableMappedIOSurfaceBitmapBackend::create):
(WebKit::ImageBufferShareableMappedIOSurfaceBitmapBackend::setOwnershipIdentity): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferShareableMappedIOSurfaceBitmapBackend.h:

Canonical link: <a href="https://commits.webkit.org/273123@main">https://commits.webkit.org/273123@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1f790ab7146af59605a67cd8924bb8b4aabe7868

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/33581 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/12354 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/35507 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/36198 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/30467 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/34626 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/14718 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/9511 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/29572 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/34056 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/10407 "Found 1 new test failure: fast/loader/redirect-to-invalid-url-using-meta-refresh-calls-policy-delegate.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/29940 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/9093 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/9185 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/29946 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/37523 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/30474 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/30278 "Found 2 new test failures: compositing/updates/animation-non-composited.html, mathml/scripts-removeChild.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/35318 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/9321 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/7273 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/33199 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/11094 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7899 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/9905 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/10088 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mattwoodrow commented on Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.cpp:
```diff
@@ -126,25 +126,27 @@ void RemoteImageBufferProxy::backingStoreWillChange()
             m_backendInfo = ImageBuffer::populateBackendInfo<ImageBufferShareableBitmapBackend>(backendParameters);
-            backend = ImageBufferShareableBitmapBackend::create(backendParameters, WTFMove(*handle));
+            auto handle = std::get<ShareableBitmap::Handle>(WTFMove(*backendHandle));
+            handle.takeOwnershipOfMemory(MemoryLedger::Graphics);
```
Isn't that a duplicate operation of the set we did in the GPU process (where we hopefully already attributed it to the web process that owns the RemoteRenderingBackend)?

kkinnunen-apple replied:
Yes, described in the commit message:

> Make some call sites of ShareableBitmap::create() more consistent.
>  The general pattern for marking the owner has to be:
> - Use ShareableBitmap::setOwnershipOfMemory() in GPUP side immediately
>   after creation.
> - Use ShareableBitmap::Handle::takeOwnershipOfMemory() in WP side
>   when receiving the handle. This is needed for the older macOS systems.

kkinnunen-apple committed:
ImageBufferShareableBitmapBackend lacks ownership attribution
https://bugs.webkit.org/show_bug.cgi?id=267266
rdar://problem/120707603

Reviewed by Matt Woodrow.

Add ShareableBitmap::setOwnershipOfMemory() to attribute the
ShareableBitmap to the owner process, currently always WP.

Use this when creating RemoteImageBuffer instances.
Use ShareableBitmap::Handle::takeOwnershipOfMemory() when receiving
the handles in WP.

Make some call sites of ShareableBitmap::create() more consistent.
The general pattern for marking the owner has to be:
- Use ShareableBitmap::setOwnershipOfMemory() in GPUP side immediately
  after creation.
- Use ShareableBitmap::Handle::takeOwnershipOfMemory() in WP side
  when receiving the handle. This is needed for the older macOS systems.

Remove unused ImageBuffer::setOwnershipIdentity() and similar methods.
Currently the ImageBuffer API contract is that the creator owns the
memory, and the internal implementation, in this case calls to
ShareableBitmap ensure it.

* Source/WebCore/platform/graphics/ImageBufferBackend.h:
(WebCore::ImageBufferBackend::setOwnershipIdentity): Deleted.
* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.cpp:
(WebCore::ImageBufferIOSurfaceBackend::transferToNewContext):
* Source/WebKit/GPUProcess/graphics/RemoteImageBuffer.cpp:
(WebKit::RemoteImageBuffer::getShareableBitmap):
(WebKit::RemoteImageBuffer::filteredNativeImage):
* Source/WebKit/Platform/SharedMemory.h:
* Source/WebKit/Shared/ShareableBitmap.h:
* Source/WebKit/Shared/ShareableBitmapHandle.cpp:
(WebKit::ShareableBitmapHandle::setOwnershipOfMemory const):
* Source/WebKit/Shared/cairo/ShareableBitmapCairo.cpp:
(WebKit::ShareableBitmap::setOwnershipIdentity):
* Source/WebKit/Shared/cg/ShareableBitmapCG.mm:
(WebKit::ShareableBitmap::setOwnershipIdentity):
* Source/WebKit/WebProcess/GPU/graphics/ImageBufferShareableBitmapBackend.cpp:
(WebKit::ImageBufferShareableBitmapBackend::create):
(WebKit::ImageBufferShareableBitmapBackend::transferToNewContext):
(WebKit::BufferShareableBitmapBackend::bytesPerRow const):
(WebKit::ImageBufferShareableBitmapBackend::bytesPerRow const): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/ImageBufferShareableBitmapBackend.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.cpp:
(WebKit::RemoteImageBufferProxy::didCreateBackend):
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferShareableMappedIOSurfaceBitmapBackend.cpp:
(WebKit::ImageBufferShareableMappedIOSurfaceBitmapBackend::create):
(WebKit::ImageBufferShareableMappedIOSurfaceBitmapBackend::setOwnershipIdentity): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferShareableMappedIOSurfaceBitmapBackend.h:

Canonical link: https://commits.webkit.org/273123@main
